src/Analyzer/Passes/FuseFunctionsPass.cpp

```diff
@@ -64,11 +85,35 @@ class FuseFunctionsVisitor : public InDepthQueryTreeVisitorWithContext<FuseFunct
             /// Do not apply for `count()` with without arguments or `count(*)`, only `count(x)` is supported.
             return;
 
-        argument_to_functions_mapping[argument_nodes[0]].insert(&node);
+        auto argument_with_source = createArgumentWithSource(argument_nodes[0]);
+        if (!argument_with_source)
+            return;
+
+        argument_to_functions_mapping[*argument_with_source].insert(&node);
     }
 
-    /// argument -> list of sum/count/avg functions with this argument
-    QueryTreeNodePtrWithHashMap<std::unordered_set<QueryTreeNodePtr *>> argument_to_functions_mapping;
```

Comment:
I think replacing this map with `QueryTreeNodePtrWithHashMap` would be enough. It will call `isEqual` and the problem will be solved. `CollectColumnSourcesVisitor` seems like overcomplication.

Reply:
It used to be like that. But it considers identical sub-queries as same trees. However, we do not want to fuse functions across sub-queries. We need here to compare columns with `isEquals`, but with strict comparison of source pointers (e.g. like in validation of group by keys)